### DISCUSSION TITLE
I hope more people can participate in the discussion of this topic !, How to render on the server without returning hydrated data next? This will increase the size and volume of HTML and document files

### DISCUSSION THREAD

**Discussion by parlay96** [vercel/next.js#74205] [Help]
### Summary

I don't want to return data here?Our company has technical experts who have achieved this step。Can we consider delaying client hydration?
![image](https://github.com/user-attachments/assets/3753850a-01d9-4391-8489-27162a74b69c)
If the page data is not returned, the document will definitely become smaller, which is beneficial for improving loading time
![image](https://github.com/user-attachments/assets/a9c8f1f0-7460-4d07-8545-79074247f30b)

### Additional information

```js
I don't want to return data here?Our company has technical experts who have achieved this step。Can we consider delaying client hydration?
![image](https://github.com/user-attachments/assets/3753850a-01d9-4391-8489-27162a74b69c)
If the page data is not returned, the document will definitely become smaller, which is beneficial for improving loading time
![image](https://github.com/user-attachments/assets/a9c8f1f0-7460-4d07-8545-79074247f30b)
```

**Comment by maciej-ka** [comment]
Based on few assumptions
(Please correct me if any of these are wrong)
- we have several components
- some of them need to be displayed immediately
- some would be better delayed
- before the delayed ones are loaded, we want to present loading placeholder

Existing solution would be to use `next/dynamic`.

### Official docs
[Lazy loading with `next/dynamic`](https://nextjs.org/docs/pages/building-your-application/optimizing/lazy-loading#nextdynamic)

### Example from the Docs
```js
import dynamic from 'next/dynamic'
 
const DynamicHeader = dynamic(() => import('../components/header'), {
  loading: () => <p>Loading...</p>,
})
 
export default function Home() {
  return <DynamicHeader />
}
```

**Reply by parlay96** [reply]
@maciej-ka My goal is how to delete __NEXT_DATA__, as it will affect the size of my HTML file

**Reply by maciej-ka** [reply]
Interesting.
- is your website 100% static?
- do you use App Router or the Pages Router?

**Comment by icyJoseph** [comment]
This is the "problem" with hydration frameworks. So called, x2 data problem. 

- https://github.com/vercel/next.js/discussions/15117#discussioncomment-36522

Here you can see, a summary of performance, and x2-ness of various frameworks: https://github.com/eknkc/ssr-benchmark

You can use this https://github.com/jherr/njs13-n-plus-one/blob/main/tester/index.js (works for App Router), tool, to measure the impact of this additional data.

The recommendation here is to massage your data within `getStaticProps`, or `getServerSideProps`, and avoid sending downstream way too much.

A consideration here is that, sensitive data, that's filtered as you render, but return from `getStaticProps`, or `getServerSideProps`, will still be present in the DOM, in `__NEXT_DATA__`.

If you choose to do a dynamic escape, you will opt that subtree out of SSR, which is not great either.

The Server Component architecture, can indeed help with this though, but it is not there just yet: https://github.com/vercel/next.js/discussions/56283#discussioncomment-7166838 - https://github.com/vercel/next.js/discussions/42170

I think this discussion was also quite good to follow: https://github.com/vercel/next.js/discussions/52119

But, yeah `__NEXT_DATA__`, as far as I remember lives on... I can tell you, most of the time it is ok, but I worked once with Akamai servers, which limit the total HTML payload, and that was no fun...

**Reply by maciej-ka** [reply]
Ton of interesting links, thanks @icyJoseph!

**Comment by parlay96** [comment]
@icyJoseph Thank you, my question is, is there any way to remove __NEXT_DATA__? Do you have any ideas? After all, it occupies a large amount of HTML and document space

**Comment by parlay96** [comment]
@timneutkens @Timer Is there any way to delete __NEXT_DATA__? Do you have any ideas? After all, it takes up a lot of HTML and document space，  Boss, can you answer that? Thank you

**Reply by icyJoseph** [reply]
https://github.com/vercel/next.js/discussions/34221

I see you have been asking this before. And again, at least in Pages Router, that's there to stay.

**Comment by parlay96** [comment]
@icyJoseph yes，But I don't know how to solve it

**Comment by parlay96** [comment]
@icyJoseph Sorry, buddy. We have already implemented the deletion

**Comment by maciej-ka** [comment]
After following your links, @icyJoseph, here is how I understand it:
_(I may got something wrong, I'm here to learn)_
- Page needs prerendered html for SEO.
- _(my conclusion)_ Next.js can't hydrate solely from html because some parts of component state may not be present there.
- So in general 2x bandwidth cannot be avoided.

Still, hydration data could arrive at different moment.
a) sent in a separate request 
a) or with component js, which @jensmeindertsma perhaps meant by [colocation](https://github.com/vercel/next.js/discussions/15117#discussioncomment-37089)

Some guessing why atm NEXT_DATA is alongside prerender:
1) To avoid mismatch: if data arrived later, there may there might be updates (e.g., DB changes, Date change), causing UI oddities.
2) Performance: hydration data shares data preparation with prerender.
3) Non-idempotency. A component that sends emails might do it twice if hydration data is fetched separately.

Could the solution be to store NEXT_DATA server side in a short-lived cache, then let the client fetch it post load?

**Reply by parlay96** [reply]
@maciej-ka I completely agree with your approach. Do you have a good implementation code

**Comment by parlay96** [comment]
@timneutkens @Timer Is there any way to delete NEXT_DATA? Do you have any ideas? After all, it takes up a lot of HTML and document space， Boss, can you answer that? Thank you

**Comment by parlay96** [comment]
I hope more people can participate in the discussion of this topic !
